dhis2/dhis2-core
# feat: cached data integrity checks [DHIS2-12599]

### Summary
On high level this adds caching to the data integrity checks.
But this has a number of consequences in the details.

The `GET` API now reads from the cache but does not trigger the checks.
To trigger the checks a `POST` is send to the same URL as reading cache.

As the computation triggered by a `POST` is now asynchronously performed by an ad-hoc scheduled job, calling `GET` directly after is likely not to find the results in the cache yet. Therefore a `timeout` parameter was added to the cache access API so that you can poll with some server side waiting (if wanted).
The idea here was that we could block wait on the cache and return as soon as the data is provided by the asynchronous computation. However, for now I did not implement block wait as the redis cache would not support it. Therefore I made a sleep wait loop trying to find it in the cache every 10-50ms (based on timeout). Since this is a very check operation I made the loop waiting relatively short to avoid long waits without a need for it.

Also the summary and details report got a `Date` field added for the `finishedTime` so that the user can understand how "old" this result is. The cache will hold results up to an hour. After that they outdate and the check must be rerun.

### Running Checks via Jobs
As now all data integrity checks are run from a job the data integrity job needed to be able to run the set of checks for one of 3 outcomes:
* report (legacy support of the old report)
* summary (new summary reports)
* details (new details reports)

This can be understood as a generally useful feature as soon as the job configuration also lets you chose this parameter in the app this means the users can also opt to run legacy or the new summary or details reports. 

For backwards compatibility the new type parameter is not required and when left out will perform the legacy report.

### Transactions
Doing all checks in  single transaction has shown problematic. While this would be nicest for consistency it has severe problems:
* a failing check can cause the transaction to fail and thereby fail all subsequent checks
* we get long running transactions potentially blocking database in significant ways (even read-only)

To avoid these problems transactions are now opened individually. For the programmatic checks this means they open when we call other services which are transnational. Usually this is the case once in these programmatic checks.
For the YAML based checks the transactions are now opened on the store level so that the service can call a spring proxy that has the `@Transactional` annotation. Therefore a new `DataIntegrityStore` and its implementation needed to be extracted from the service.

### Check Bugfix
The first YAML check had bad SQL for the summary check.
The convention is that a summary should return count and possibly a percentage.
The SQL was copied as still contained some text columns added for readability when used directly.  

### Automatic Testing
Existing tests have been adopted to cover the new asynchronous processing.
Luckily this was already setup for the legacy reports.

### Manual Testing
* Start the server and login
* Trigger summary: `POST` http://localhost:8080/api/dataIntegrity/summary
* Check summary: `GET` http://localhost:8080/api/dataIntegrity/summary?timeout=500
* Trigger details: `POST` http://localhost:8080/api/dataIntegrity/details
* Check details: `GET` http://localhost:8080/api/dataIntegrity/details?timeout=500
 
Note that a timeout of 500ms may not be long enough when running all checks like the examples given will do.
You could also test just single checks like:
* Trigger: `POST` http://localhost:8080/api/dataIntegrity/summary?checks=categories-no-options
* Check: `GET` http://localhost:8080/api/dataIntegrity/summary?checks=categories-no-options?timeout=500

### Documentation
https://github.com/dhis2/dhis2-docs/pull/906
